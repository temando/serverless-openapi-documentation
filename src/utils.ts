import { Clone, IMerge, Merge } from 'lutils';

export const merge: IMerge = new Merge({ depth: 100 }).merge;
export const clone = new Clone({ depth: 100 }).clone;

export function isIterable (obj) {
  if (obj === null || obj === undefined) {
    return false;
  }

  return typeof obj[Symbol.iterator] === 'function';
}

export function omit<T extends object> (obj: T, keys: string[]): T {
  const cloned = clone(obj);

  for (const key of keys) {
    delete cloned[key];
  }

  return cloned;
}
