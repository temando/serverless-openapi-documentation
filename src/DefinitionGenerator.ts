import { dereference } from '@jdw/jst';
// tslint:disable-next-line no-submodule-imports
import { validateSync as openApiValidatorSync } from 'swagger2openapi/validate';
import * as uuid from 'uuid';
import { IDefinition, IDefinitionConfig, IOperation, IParameterConfig, IServerlessFunctionConfig } from './types';
import { clone, isIterable, merge } from './utils';

export class DefinitionGenerator {
  // The OpenAPI version we currently validate against
  public version = '3.0.0';

  // Base configuration object
  public definition = <IDefinition> {
    openapi: this.version,
    components: {},
  };

  public config: IDefinitionConfig;

  /**
   * Constructor
   * @param serviceDescriptor IServiceDescription
   */
  constructor (config: IDefinitionConfig) {
    this.config = clone(config);
  }

  public parse () {
    const {
      title = '',
      description = '',
      version = uuid.v4(),
      models,
    } = this.config;

    merge(this.definition, {
      openapi: this.version,
      info: { title, description, version },
      paths: {},
      components: {
        schemas: {},
        securitySchemes: {},
      },
    });

    if (isIterable(models)) {
      for (const model of models) {
        if (!model.schema) {
          continue;
        }

        this.definition.components.schemas[model.name] = this.cleanSchema(
          dereference(model.schema),
        );
      }
    }

    return this;
  }

  public validate (): { valid: boolean, context: string[], warnings: any[], error?: any[] } {
    const payload: any = {};

    try {
      openApiValidatorSync(this.definition, payload);
    } catch (error) {
      payload.error = error.message;
    }

    return payload;
  }

  /**
   * Add Paths to OpenAPI Configuration from Serverless function documentation
   * @param config Add
   */
  public readFunctions (config: IServerlessFunctionConfig[]): void {
    // loop through function configurations
    for (const funcConfig of config) {
      // loop through http events
      for (const httpEvent of this.getHttpEvents(funcConfig.events)) {
        const httpEventConfig = httpEvent.http;

        if (httpEventConfig.documentation) {
          // Build OpenAPI path configuration structure for each method
          const pathConfig = {
            [`/${httpEventConfig.path}`]: {
              [httpEventConfig.method.toLowerCase()]: this.getOperationFromConfig(
                funcConfig._functionName,
                httpEventConfig.documentation,
              ),
            },
          };

          // merge path configuration into main configuration
          merge(this.definition.paths, pathConfig);
        }
      }
    }
  }

  /**
   * Cleans schema objects to make them OpenAPI compatible
   * @param schema JSON Schema Object
   */
  private cleanSchema (schema) {
    // Clone the schema for manipulation
    const cleanedSchema = clone(schema);

    // Strip $schema from schemas
    if (cleanedSchema.$schema) {
      delete cleanedSchema.$schema;
    }

    // Return the cleaned schema
    return cleanedSchema;
  }

  /**
   * Generate Operation objects from the Serverless Config.
   *
   * @link https://github.com/OAI/OpenAPI-Specification/blob/3.0.0/versions/3.0.0.md#operationObject
   * @param funcName
   * @param documentationConfig
   */
  private getOperationFromConfig (funcName: string, documentationConfig): IOperation {
    const operationObj: IOperation = {
      operationId: funcName,
    };

    if (documentationConfig.summary) {
      operationObj.summary = documentationConfig.summary;
    }

    if (documentationConfig.description) {
      operationObj.description = documentationConfig.description;
    }

    if (documentationConfig.tags) {
      operationObj.tags = documentationConfig.tags;
    }

    if (documentationConfig.deprecated) {
      operationObj.deprecated = true;
    }

    if (documentationConfig.requestBody) {
      operationObj.requestBody = this.getRequestBodiesFromConfig(documentationConfig);
    }

    if (documentationConfig.parameters) {
      operationObj.parameters = this.getParametersFromConfig(documentationConfig);
    }

    operationObj.responses = this.getResponsesFromConfig(documentationConfig);

    return operationObj;
  }

  /**
   * Derives Path, Query and Request header parameters from Serverless documentation
   * @param documentationConfig
   */
  private getParametersFromConfig (documentationConfig): IParameterConfig[] {
    const parameters: IParameterConfig[] = [];

    // Build up parameters from configuration for each parameter type
    for (const type of ['path', 'query', 'header', 'cookie']) {
      let paramBlock;
      if (type === 'path' && documentationConfig.pathParams) {
        paramBlock = documentationConfig.pathParams;
      } else if (type === 'query' && documentationConfig.queryParams) {
        paramBlock = documentationConfig.queryParams;
      } else if (type === 'header' && documentationConfig.requestHeaders) {
        paramBlock = documentationConfig.requestHeaders;
      } else if (type === 'cookie' && documentationConfig.cookieParams) {
        paramBlock = documentationConfig.cookieParams;
      } else {
        continue;
      }

      // Loop through each parameter in a parameter block and add parameters to array
      for (const parameter of paramBlock) {
        const parameterConfig: IParameterConfig = {
          name: parameter.name,
          in: type,
          description: parameter.description || '',
          required: parameter.required || false, // Note: all path parameters must be required
        };

        // if type is path, then required must be true (@see OpenAPI 3.0-RC1)
        if (type === 'path') {
          parameterConfig.required = true;
        } else if (type === 'query') {
          parameterConfig.allowEmptyValue = parameter.allowEmptyValue || false;  // OpenAPI default is false

          if ('allowReserved' in parameter) {
            parameterConfig.allowReserved = parameter.allowReserved || false;
          }
        }

        if ('deprecated' in parameter) {
          parameterConfig.deprecated = parameter.deprecated;
        }

        if ('style' in parameter) {
          parameterConfig.style = parameter.style;

          parameterConfig.explode = parameter.explode
            ? parameter.explode
            : parameter.style === 'form';
        }

        if (parameter.schema) {
          parameterConfig.schema = this.cleanSchema(parameter.schema);
        }

        if (parameter.example) {
          parameterConfig.example = parameter.example;
        } else if (parameter.examples && Array.isArray(parameter.examples)) {
          parameterConfig.examples = parameter.examples;
        }

        if (parameter.content) {
          parameterConfig.content = parameter.content;
        }

        parameters.push(parameterConfig);
      }
    }

    return parameters;
  }

  /**
   * Derives request body schemas from event documentation configuration
   * @param documentationConfig
   */
  private getRequestBodiesFromConfig (documentationConfig) {
    const requestBodies = {};

    if (!documentationConfig.requestModels) {
      throw new Error(`Required requestModels in: ${JSON.stringify(documentationConfig, null, 2)}`);
    }

    // Does this event have a request model?
    if (documentationConfig.requestModels) {
      // For each request model type (Sorted by "Content-Type")
      for (const requestModelType of Object.keys(documentationConfig.requestModels)) {
        // get schema reference information
        const requestModel = this.config.models.filter(
          (model) => model.name === documentationConfig.requestModels[requestModelType],
        ).pop();

        if (requestModel) {
          const reqModelConfig = {
            schema: {
              $ref: `#/components/schemas/${documentationConfig.requestModels[requestModelType]}`,
            },
          };

          this.attachExamples(requestModel, reqModelConfig);

          const reqBodyConfig: { content: object, description?: string } = {
            content: {
              [requestModelType]: reqModelConfig,
            },
          };

          if (documentationConfig.requestBody && 'description' in documentationConfig.requestBody) {
            reqBodyConfig.description = documentationConfig.requestBody.description;
          }

          merge(requestBodies, reqBodyConfig);
        }
      }
    }

    return requestBodies;
  }

  private attachExamples (target, config) {
    if (target.examples && Array.isArray(target.examples)) {
      merge(config, { examples: clone(target.examples) });
    } else if (target.example) {
      merge(config, { example: clone(target.example) });
    }
  }

  /**
   * Gets response bodies from documentation config
   * @param documentationConfig
   */
  private getResponsesFromConfig (documentationConfig) {
    const responses = {};
    if (documentationConfig.methodResponses) {
      for (const response of documentationConfig.methodResponses) {
        const methodResponseConfig: { description: any, content: object, headers?: object } = {
          description: (
            (response.responseBody && 'description' in response.responseBody)
              ? response.responseBody.description
              : `Status ${response.statusCode} Response`
          ),
          content: this.getResponseContent(response.responseModels),
        };

        if (response.responseHeaders) {
          methodResponseConfig.headers = {};
          for (const header of response.responseHeaders) {
            methodResponseConfig.headers[header.name] = {
              description: header.description || `${header.name} header`,
            };
            if (header.schema) {
              methodResponseConfig.headers[header.name].schema = this.cleanSchema(header.schema);
            }
          }
        }

        merge(responses, {
          [response.statusCode]: methodResponseConfig,
        });
      }
    }

    return responses;
  }

  private getResponseContent (response) {
    const content = {};

    for (const responseKey of Object.keys(response)) {
      const responseModel = this.config.models.find((model) =>
        model.name === response[responseKey],
      );

      if (responseModel) {
        const resModelConfig = {
          schema: {
            $ref: `#/components/schemas/${response[responseKey]}`,
          },
        };

        this.attachExamples(responseModel, resModelConfig);

        merge(content, { [responseKey] : resModelConfig });
      }
    }

    return content;
  }

  private getHttpEvents (funcConfig) {
    return funcConfig.filter((event) => event.http ? true : false);
  }
}
