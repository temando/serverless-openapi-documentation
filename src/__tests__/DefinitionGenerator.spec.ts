import * as path from 'path';
import * as Serverless from 'serverless';
import { DefinitionGenerator } from '../DefinitionGenerator';
import { merge } from '../utils';

class ServerlessInterface extends Serverless {
  public service: any = {};
  public config: any = {};
  public yamlParser: any = {};
  public pluginManager: any = {};
  public variables: any = {};
}

describe('OpenAPI Documentation Generator', () => {
  it('Generates OpenAPI document', async () => {
    const servicePath = path.join(__dirname, '../../test/project');
    const serverlessYamlPath = path.join(servicePath, './serverless.yml');
    const sls: ServerlessInterface = new Serverless();

    sls.config.update({
      servicePath,
    });

    const config = await sls.yamlParser.parse(serverlessYamlPath);
    expect(config).not.toBeNull();
    sls.pluginManager.cliOptions = { stage: 'dev' };

    await sls.service.load(config);
    await sls.variables.populateService();

    if ('documentation' in sls.service.custom) {
      const docGen = new DefinitionGenerator(sls.service.custom.documentation);
      docGen.parse();
       // Map function configurations
      const funcConfigs = sls.service.getAllFunctions().map((functionName) => {
        const func = sls.service.getFunction(functionName);
        return merge({ _functionName: functionName }, func);
      });

      // Add Paths to OpenAPI Output from Function Configuration
      docGen.readFunctions(funcConfigs);

      expect(docGen.definition).not.toBeNull();
      expect(docGen.definition).toMatchSnapshot();
    } else {
      throw new Error('Cannot find "documentation" in custom section of "serverless.yml"');
    }
  });
});
