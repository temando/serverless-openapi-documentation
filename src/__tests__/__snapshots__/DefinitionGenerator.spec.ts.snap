// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenAPI Documentation Generator Generates OpenAPI document 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "ErrorResponse": Object {
        "additionalProperties": false,
        "description": "This is a schema for responses in the JSON API format. For more, see http://jsonapi.org",
        "properties": Object {
          "errors": Object {
            "items": Object {
              "additionalProperties": false,
              "properties": Object {
                "code": Object {
                  "description": "An application-specific error code, expressed as a string value.",
                  "type": "string",
                },
                "detail": Object {
                  "description": "A human-readable explanation specific to this occurrence of the problem.",
                  "type": "string",
                },
                "id": Object {
                  "description": "A unique identifier for this particular occurrence of the problem.",
                  "type": "string",
                },
                "links": Object {
                  "additionalProperties": true,
                  "description": "A resource object **MAY** contain references to other resource objects (\\"relationships\\"). Relationships may be to-one or to-many. Relationships can be specified by including a member in a resource's links object.",
                  "properties": Object {
                    "related": Object {
                      "description": "A link **MUST** be represented as either: a string containing the link's URL or a link object.",
                      "oneOf": Array [
                        Object {
                          "description": "A string containing the link's URL.",
                          "format": "uri",
                          "type": "string",
                        },
                        Object {
                          "properties": Object {
                            "href": Object {
                              "description": "A string containing the link's URL.",
                              "format": "uri",
                              "type": "string",
                            },
                            "meta": Object {
                              "additionalProperties": true,
                              "description": "Non-standard meta-information that can not be represented as an attribute or relationship.",
                              "type": "object",
                            },
                          },
                          "required": Array [
                            "href",
                          ],
                          "type": "object",
                        },
                      ],
                    },
                    "self": Object {
                      "description": "A \`self\` member, whose value is a URL for the relationship itself (a \\"relationship URL\\"). This URL allows the client to directly manipulate the relationship. For example, it would allow a client to remove an \`author\` from an \`article\` without deleting the people resource itself.",
                      "format": "uri",
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
                "meta": Object {
                  "additionalProperties": true,
                  "description": "Non-standard meta-information that can not be represented as an attribute or relationship.",
                  "type": "object",
                },
                "source": Object {
                  "properties": Object {
                    "parameter": Object {
                      "description": "A string indicating which query parameter caused the error.",
                      "type": "string",
                    },
                    "pointer": Object {
                      "description": "A JSON Pointer [RFC6901] to the associated entity in the request document [e.g. \\"/data\\" for a primary data object, or \\"/data/attributes/title\\" for a specific attribute].",
                      "type": "string",
                    },
                  },
                  "type": "object",
                },
                "status": Object {
                  "description": "The HTTP status code applicable to this problem, expressed as a string value.",
                  "type": "string",
                },
                "title": Object {
                  "description": "A short, human-readable summary of the problem. It **SHOULD NOT** change from occurrence to occurrence of the problem, except for purposes of localization.",
                  "type": "string",
                },
              },
              "type": "object",
            },
            "type": "array",
            "uniqueItems": true,
          },
          "links": Object {
            "additionalProperties": true,
            "description": "A resource object **MAY** contain references to other resource objects (\\"relationships\\"). Relationships may be to-one or to-many. Relationships can be specified by including a member in a resource's links object.",
            "properties": Object {
              "related": Object {
                "description": "A link **MUST** be represented as either: a string containing the link's URL or a link object.",
                "oneOf": Array [
                  Object {
                    "description": "A string containing the link's URL.",
                    "format": "uri",
                    "type": "string",
                  },
                  Object {
                    "properties": Object {
                      "href": Object {
                        "description": "A string containing the link's URL.",
                        "format": "uri",
                        "type": "string",
                      },
                      "meta": Object {
                        "additionalProperties": true,
                        "description": "Non-standard meta-information that can not be represented as an attribute or relationship.",
                        "type": "object",
                      },
                    },
                    "required": Array [
                      "href",
                    ],
                    "type": "object",
                  },
                ],
              },
              "self": Object {
                "description": "A \`self\` member, whose value is a URL for the relationship itself (a \\"relationship URL\\"). This URL allows the client to directly manipulate the relationship. For example, it would allow a client to remove an \`author\` from an \`article\` without deleting the people resource itself.",
                "format": "uri",
                "type": "string",
              },
            },
            "type": "object",
          },
          "meta": Object {
            "additionalProperties": true,
            "description": "Non-standard meta-information that can not be represented as an attribute or relationship.",
            "type": "object",
          },
        },
        "required": Array [
          "errors",
        ],
        "title": "JSON API Schema",
        "type": "object",
      },
      "PutDocumentRequest": Object {
        "properties": Object {
          "SomeObject": Object {
            "properties": Object {
              "SomeAttribute": Object {
                "type": "string",
              },
            },
            "type": "object",
          },
        },
      },
      "PutDocumentResponse": Object {
        "title": "Empty Schema",
        "type": "object",
      },
    },
    "securitySchemes": Object {
      "auth": Object {
        "scheme": "bearer",
        "type": "http",
      },
    },
  },
  "info": Object {
    "description": "a description",
    "title": "a title",
    "version": "1.2",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/create": Object {
      "post": Object {
        "description": "Creates a user and then sends a generated password email",
        "operationId": "createUser",
        "parameters": Array [
          Object {
            "description": "The username for a user to create",
            "in": "path",
            "name": "username",
            "required": true,
            "schema": Object {
              "pattern": "^[-a-z0-9_]+$",
              "type": "string",
            },
          },
          Object {
            "allowEmptyValue": false,
            "description": "The user's Membership Type",
            "in": "query",
            "name": "membershipType",
            "required": false,
            "schema": Object {
              "enum": Array [
                "premium",
                "standard",
              ],
              "type": "string",
            },
          },
          Object {
            "description": "A Session ID variable",
            "in": "cookie",
            "name": "SessionId",
            "required": false,
            "schema": Object {
              "type": "string",
            },
          },
        ],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$ref": "#/components/schemas/PutDocumentRequest",
              },
            },
          },
          "description": "A user information object",
        },
        "responses": Object {
          "201": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/PutDocumentResponse",
                },
              },
            },
            "description": "A user object along with generated API Keys",
          },
          "500": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/ErrorResponse",
                },
              },
            },
            "description": "An error message when creating a new user",
          },
        },
        "security": Array [
          Object {
            "auth": Array [],
          },
        ],
        "summary": "Create User",
      },
    },
  },
  "servers": Array [
    Object {
      "description": "Dev server",
      "url": "https://example.com",
    },
  ],
}
`;
